sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -47,9 +47,14 @@ object HiveTypeCoercion {
     case (t1, NullType) => Some(t1)
 
     // Promote numeric types to the highest of the two and all numeric types to unlimited decimal
+    // Promote numeric types to String if one is string type and the other is numeric type
     case (t1, t2) if Seq(t1, t2).forall(numericPrecedence.contains) =>
```

Comment:
`numericPrecedence` doesn't contain `StringType`, you need to write a separate case like `case (t1 @ StringType(), t2 @ NumericType())`.